File: module-quote.1/MyQuote/app/src/main/java/com/developer/cookie/myquote/quote/abstract_class/SingleFragmentAbstractActivity.java
package com.developer.cookie.myquote.quote.abstract_class;


import android.content.Intent;
import android.content.res.Configuration;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;

import com.developer.cookie.myquote.R;
import com.developer.cookie.myquote.quote.activities.AddQuoteActivity;
import com.developer.cookie.myquote.utils.AppBarLayoutExpended;
import com.developer.cookie.myquote.utils.ColorationTextChar;

import java.util.Locale;

/**
 * SingleFragmentAbstractActivity helps avoid boilerplate code.
 */
public abstract class SingleFragmentAbstractActivity extends AppCompatActivity {
    public Fragment currentFragment;
    public FloatingActionButton fab;
    public int fabImageResourceId;
    public String mQuotesType;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        defineInputData(savedInstanceState);
        setContentView(getLayoutResId());
        defineToolbar();
        setAppBarNotExpandable();
        setNewTitleStyle(mQuotesType);
        defineFragment();
        defineFab();
    }

    public void defineInputData(Bundle saveInstanceState) { }

    @LayoutRes
    protected int getLayoutResId() {
        return R.layout.activity_single_fragment;
    }

    private void defineToolbar() {
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
    }

    private void setAppBarNotExpandable() {
        if (findViewById(R.id.detail_fragment_container) == null) {
            AppBarLayout appBarLayout = (AppBarLayout) findViewById(R.id.appbar_layout);
            CoordinatorLayout.LayoutParams layoutParams = (CoordinatorLayout.LayoutParams) appBarLayout.getLayoutParams();
            CollapsingToolbarLayout collapsingToolbarLayout = (CollapsingToolbarLayout) findViewById(R.id.collapsing_toolbar);
            Configuration configuration = getResources().getConfiguration();
            AppBarLayoutExpended.setAppBarLayoutExpended(this, appBarLayout, layoutParams, collapsingToolbarLayout, configuration);
        }
    }

    public void setNewTitleStyle(String quoteType) {
        String localeLanguage = Locale.getDefault().getLanguage();
        setTitle(ColorationTextChar.setFirstVowelColor(quoteType, localeLanguage, this));
    }

    private void defineFragment() {
        FragmentManager fragmentManager = getSupportFragmentManager();
        currentFragment = fragmentManager.findFragmentById(R.id.container);
        if (currentFragment == null) {
            currentFragment = createFragment();
            fragmentManager.beginTransaction()
                    .add(R.id.container, currentFragment)
                    .commit();
        }
    }

    public abstract Fragment createFragment();

    private void defineFab() {
        fabImageResourceId = setFabImageResourceId();
        fab = (FloatingActionButton) findViewById(R.id.fab);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            fab.setImageDrawable(getResources().getDrawable(fabImageResourceId, getTheme()));
        } else {
            fab.setImageDrawable(getResources().getDrawable(fabImageResourceId));
        }
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                defineActionWhenFabIsPressed();
            }
        });
    }

    public int setFabImageResourceId() {
        return R.drawable.ic_add_white_24dp;
    }

    public void defineActionWhenFabIsPressed() {
        Intent intent = AddQuoteActivity.newIntent(this, getTitle().toString());
        startActivity(intent);
    }
}

File: app/src/main/java/com/github/veselinazatchepina/myquotes/abstracts/SingleFragmentAbstractActivity.kt
package com.github.veselinazatchepina.myquotes.abstracts

import android.content.res.Configuration
import android.os.Build
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.view.animation.OvershootInterpolator
import com.github.veselinazatchepina.myquotes.R
import com.github.veselinazatchepina.myquotes.addquote.AddQuoteActivity
import com.github.veselinazatchepina.myquotes.enums.QuoteType
import com.github.veselinazatchepina.myquotes.setFirstVowelColor
import icepick.Icepick
import kotlinx.android.synthetic.main.activity_single_fragment.*
import kotlinx.android.synthetic.main.fab_popup_menu.*


abstract class SingleFragmentAbstractActivity : AppCompatActivity() {

    private lateinit var fabOpenAnimation: Animation
    private lateinit var fabCloseAnimation: Animation
    private var isFabOpen = false

    companion object {
        const val FRAGMENT_TAG = "fragment_tag"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        setTheme(R.style.AppTheme)
        super.onCreate(savedInstanceState)
        Icepick.restoreInstanceState(this, savedInstanceState)
        setContentView(getLayoutResId())
        defineInputData()
        defineToolbar()
        defineNavigationDrawer()
        defineAppBarLayoutExpandableValue()
        defineFab()
        if (title != null) {
            setNewTitleStyle(title.toString())
        }
        defineFragment()
        createPresenter()
    }

    open fun getLayoutResId(): Int = R.layout.activity_current_single_fragment

    open fun defineInputData() {}

    private fun defineToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    open fun defineNavigationDrawer() {}

    open fun defineAppBarLayoutExpandableValue() {
        setAppBarNotExpandable()
    }

    fun setAppBarNotExpandable() {
        appbarLayout.setExpanded(false, false)
        if (resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) {
            appbarLayout.layoutParams.height = this.resources
                    .getDimension(R.dimen.toolbar_height_normal_portrait).toInt()
        } else {
            appbarLayout.layoutParams.height = this.resources
                    .getDimension(R.dimen.toolbar_height_normal_landscape).toInt()
        }
        collapsingToolbar.isTitleEnabled = false
    }

    private fun defineFab() {
        defineFabAnimation()
        val fabImageResourceId = setFabImageResId()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            addFab.setImageDrawable(resources.getDrawable(fabImageResourceId, theme))
        } else {
            addFab.setImageDrawable(resources.getDrawable(fabImageResourceId))
        }
        defineActionWhenFabIsPressed()
    }

    private fun defineFabAnimation() {
        fabOpenAnimation = AnimationUtils.loadAnimation(this, R.anim.open_fab_menu)
        fabCloseAnimation = AnimationUtils.loadAnimation(this, R.anim.close_fab_menu)
    }

    open fun setFabImageResId(): Int = R.drawable.ic_add_white_24dp

    open fun defineActionWhenFabIsPressed() {
        addFab.setOnClickListener {
            defineBookFabListener()
            defineMyQuoteFabListener()
            if (!isFabOpen) {
                showFabMenu()
            } else {
                closeFabMenu()
            }
        }
    }

    private fun defineBookFabListener() {
        bookQuoteFab.setOnClickListener {
            startActivity(AddQuoteActivity.newIntent(this, getString(QuoteType.BOOK_QUOTE.resource)))
            closeFabMenu()
        }
    }

    private fun defineMyQuoteFabListener() {
        myQuoteFab.setOnClickListener {
            startActivity(AddQuoteActivity.newIntent(this, getString(QuoteType.MY_QUOTE.resource)))
            closeFabMenu()
        }
    }

    private fun showFabMenu() {
        addFab.animate().rotation(45.0F)
                .withLayer()
                .setDuration(300)
                .setInterpolator(OvershootInterpolator(10.0F))
                .start()
        bookQuoteFab.startAnimation(fabOpenAnimation)
        myQuoteFab.startAnimation(fabOpenAnimation)
        isFabOpen = true
    }

    private fun closeFabMenu() {
        addFab.animate().rotation(0.0F)
                .withLayer()
                .setDuration(300)
                .setInterpolator(OvershootInterpolator(10.0F))
                .start()
        bookQuoteFab.startAnimation(fabCloseAnimation)
        myQuoteFab.startAnimation(fabCloseAnimation)
        isFabOpen = false
    }

    open fun setNewTitleStyle(title: String) {
        setTitle(title.setFirstVowelColor(this))
    }

    private fun defineFragment() {
        var currentFragment = supportFragmentManager.findFragmentById(R.id.container)
        if (currentFragment == null) {
            currentFragment = createFragment()
            supportFragmentManager.beginTransaction()
                    .add(R.id.container, currentFragment, FRAGMENT_TAG)
                    .commit()
        }
    }

    abstract fun createFragment(): Fragment

    abstract fun createPresenter()

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        Icepick.saveInstanceState(this, outState)
    }
}